packages/backend-core/src/features/index.ts

```diff
@@ -1,79 +1,108 @@
-    })
-  })
+const DEFAULTS = Object.keys(FLAGS).reduce((acc, key) => {
+  const typedKey = key as keyof typeof FLAGS
+  // @ts-ignore
+  acc[typedKey] = FLAGS[typedKey].defaultValue
+  return acc
+}, {} as Flags)
 
-  return tenantFeatureFlags
+type UnwrapFlag<F> = F extends Flag<infer U> ? U : never
+export type Flags = {
+  [K in keyof typeof FLAGS]: UnwrapFlag<(typeof FLAGS)[K]>
 }
 
-export function isEnabled(featureFlag: string) {
-  const tenantId = context.getTenantId()
-  const flags = getTenantFeatureFlags(tenantId)
-  return flags.includes(featureFlag)
+// Exported for use in tests, should not be used outside of this file.
+export function defaultFlags(): Flags {
+  return cloneDeep(DEFAULTS)
 }
 
-export function getTenantFeatureFlags(tenantId: string) {
-  let flags: string[] = []
-  const envFlags = buildFeatureFlags()
-  if (envFlags) {
-    const globalFlags = envFlags["*"]
-    const tenantFlags = envFlags[tenantId] || []
-
-    // Explicitly exclude tenants from global features if required.
-    // Prefix the tenant flag with '!'
-    const tenantOverrides = tenantFlags.reduce(
-      (acc: string[], flag: string) => {
-        if (flag.startsWith("!")) {
-          let stripped = flag.substring(1)
-          acc.push(stripped)
-        }
-        return acc
-      },
-      []
-    )
-
-    if (globalFlags) {
-      flags.push(...globalFlags)
-    }
-    if (tenantFlags.length) {
-      flags.push(...tenantFlags)
+function isFlagName(name: string): name is keyof Flags {
+  return FLAGS[name as keyof typeof FLAGS] !== undefined
+}
+
+/**
+ * Reads the TENANT_FEATURE_FLAGS environment variable and returns a Flags object
+ * populated with the flags for the current tenant, filling in the default values
+ * if the flag is not set.
+ *
+ * Check the tests for examples of how TENANT_FEATURE_FLAGS should be formatted.
+ *
+ * In future we plan to add more ways of setting feature flags, e.g. PostHog, and
+ * they will be accessed through this function as well.
+ */
+export async function fetch(): Promise<Flags> {
+  const currentTenantId = context.getTenantId()
+  const flags = defaultFlags()
+
+  const split = (env.TENANT_FEATURE_FLAGS || "")
+    .split(",")
+    .map(x => x.split(":"))
+  for (const [tenantId, ...features] of split) {
+    if (!tenantId || (tenantId !== "*" && tenantId !== currentTenantId)) {
+      continue
     }
 
-    // Purge any tenant specific overrides
-    flags = flags.filter(flag => {
-      return tenantOverrides.indexOf(flag) == -1 && !flag.startsWith("!")
-    })
+    for (let feature of features) {
+      let value = true
+      if (feature.startsWith("!")) {
+        feature = feature.slice(1)
+        value = false
+      }
+
+      if (!isFlagName(feature)) {
+        throw new Error(`Feature: ${feature} is not an allowed option`)
+      }
+
+      if (typeof flags[feature] !== "boolean") {
+        throw new Error(`Feature: ${feature} is not a boolean`)
+      }
```

Comment:
I was torn between making these errors or making them log a warning and get ignored. Because they're an environment variable, I erred toward erroring. It's reasonable to expect the environment variable to be strictly correct, imo.

When I come to add PostHog to this, I _think_ it makes more sense to ignore unknown flags. This will let us set them up before use and we wouldn't want adding a new flag in the PostHog UI to suddenly crash all of Budicloud.

Reply:
Yeah I think I would generally tend towards ignoring anything we don't understand rather than error - while its good to enforce strictness here, if any of this is linked up to posthog there will be a risk of a catastrophic error, when it probably would be better to ignore and log the issue.

Reply:
The idea would be to ignore missing flags for PostHog and error here for the environment variables. If you'd prefer I ignore for the environment variables it's easy to change, though I think it's better to error.